### PR TITLE
Add timer and fix soft start construction parameter order bug in robot

### PR DESCRIPTION
**Pull request by jdcaporale**
This commit adds a realtime timer to remove the cycle_count that measure update
loops and fixes a construction bug where torque and duration were flipped.
In the case that the cycle time overflowed the intended frequency the
cycle count is no longer meaningful and gives the user little understanding of
the time past. A timer is agnostic to a misbehaving update cycle.

### CODE REVIEW

File: include/kodlab_mjbots_sdk/robot_base.h [hunk 1]
```diff
@@ -82,7 +82,7 @@ namespace kodlab
          * @warning All derivative classes overriding this method should include
          *          a cycle count increment (i.e., `cycle_count_++;`).
          */
-        virtual void Update(){cycle_count_++;}; //TODO remove cycle_count and use time or more intelligently set cycle_count or properly handle softstart
+        virtual void Update(){}; //TODO remove cycle_count and use time or more intelligently set cycle_count or properly handle softstart
```
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/robot_base.h:85]:
```suggestion
        virtual void Update(){}; 
```

File: src/soft_start.cpp [hunk 2]
```diff
@@ -17,16 +17,15 @@ float SoftStart::Constrain(float values, float min_val, float max_val) {
...
     Constrain(torques, -max_val, max_val);
-
   }
 
 }
-SoftStart::SoftStart(float max_torque, int duration)
-    : max_torque_(fabs(max_torque)), duration_(duration), slope_(fabs(max_torque) / (fmax(duration, 1))) {}
+SoftStart::SoftStart(float max_torque, float duration_ms)
+    : max_torque_(fabs(max_torque)), duration_ms_(duration_ms), slope_(fabs(max_torque) / (fmax(duration_ms, 1.0))) {}
 } // namespace kodlab
```
Comment (jdcaporale) [inline src/soft_start.cpp:31]:
```suggestion
} // namespace kodlab

```

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h [hunk 3]
```diff
@@ -166,7 +166,7 @@ MjbotsControlLoop<log_type, input_type, robot_type>::MjbotsControlLoop(std::vect
 template<class log_type, class input_type, class robot_type>
 MjbotsControlLoop<log_type, input_type, robot_type>::MjbotsControlLoop(
     std::vector<std::shared_ptr<kodlab::mjbots::JointMoteus>> joint_ptrs, const ControlLoopOptions &options)
-  : MjbotsControlLoop(std::make_shared<robot_type>(joint_ptrs,options.soft_start_duration, options.max_torque),options) {}
+  : MjbotsControlLoop(std::make_shared<robot_type>(joint_ptrs, options.max_torque, options.soft_start_duration_ms),options) {}
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/mjbots_control_loop.h:169]: Can't believe we're only just seeing this, good catch!

**Comment by jdcaporale** [comment]
#37

**Review by ethanmusser — APPROVED** [review]
Looks good other than the two minor issues that you already pointed out.

File: include/kodlab_mjbots_sdk/soft_start.h [hunk 4]
```diff
@@ -16,7 +16,7 @@ class SoftStart {
  private:
   float slope_ = 0;        /// Slope of the ramp
   float max_torque_ = 100; /// Max torque allowed in Nm
-  int duration_ = 0;       /// Duration of soft Start in steps
+  float duration_ms_ = 0;   /// Duration of soft Start in steps
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/soft_start.h:19]: fix comment

File: include/kodlab_mjbots_sdk/soft_start.h [hunk 5]
```diff
@@ -25,7 +25,7 @@ class SoftStart {
    * @param torques[in, out] the torques to be constrained
    * @param count the number of steps it has been
    */
-  void ConstrainTorques(std::vector<float> &torques, int count);
+  void ConstrainTorques(std::vector<float> &torques, float time_since_start_ms);
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/soft_start.h:28]: fix comment

File: src/soft_start.cpp [hunk 6]
```diff
@@ -17,16 +17,15 @@ float SoftStart::Constrain(float values, float min_val, float max_val) {
   return fmin(fmax(values, min_val), max_val);
 }
 
-void SoftStart::ConstrainTorques(std::vector<float> &torques, int count) {
-  if (count > duration_) {
+void SoftStart::ConstrainTorques(std::vector<float> &torques, float time_since_start_ms) {
+  if ((time_since_start_ms) > duration_ms_) { // ramp_timer returns in us
```
Comment (ShaneRozenLevy) [inline src/soft_start.cpp:21]: comment does not make sense

**Review by ShaneRozenLevy — APPROVED** [review]
Some comment issues, but change is fine.

File: include/kodlab_mjbots_sdk/robot_base.h (hunk 1 above)
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/robot_base.h]: fixed in 21511f7a0d652fe35a9fc578d4b151ac0b83c9d4

File: src/soft_start.cpp (hunk 2 above)
Comment (ethanmusser) [reply src/soft_start.cpp]: Fixed in 21511f7a0d652fe35a9fc578d4b151ac0b83c9d4

File: include/kodlab_mjbots_sdk/soft_start.h (hunk 4 above)
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/soft_start.h]: Fixed in 21511f7a0d652fe35a9fc578d4b151ac0b83c9d4

File: include/kodlab_mjbots_sdk/soft_start.h (hunk 5 above)
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/soft_start.h:30]: Fixed in 21511f7a0d652fe35a9fc578d4b151ac0b83c9d4

File: src/soft_start.cpp (hunk 6 above)
Comment (ethanmusser) [reply src/soft_start.cpp]: Fixed in 21511f7a0d652fe35a9fc578d4b151ac0b83c9d4

**Pull request merged**